### PR TITLE
Fixed bug in regex of std.code.member.fields now more variables are found

### PR DESCRIPTION
**Pull request by noricor**
Hi, 

when trying out option --std.code.member.fields I found that almost no variables are detected.

The original problem was a missing * in the regular expression
```python
r'''([_a-zA-Z][_a-zA-Z0-9]*\s+[_a-zA-Z][_a-zA-Z0-9])\s*[=;]''')
```
the expression should instead be
```python
r'''([_a-zA-Z][_a-zA-Z0-9]*\s+[_a-zA-Z][_a-zA-Z0-9]*)\s*[=;]''')
```

After implementing the change still many variables were not detected, so I came up with 
a more sophisticated expression for cpp which finds much more but may still not be complete
(cpp syntax is a nightmare to match)

With this improvement the tool now can find stuff like this:
```cpp
class Test
{
    int a;
    long int b;
    char const* c;
    extern const int& d;
    int e = 0;
    const int f = 42;
    int gg;
    std::string m_string;
    static ::std::vector<std::function<int(void)>> const& m_vect;
    Foo::Bar::bla m_gar;

    static std::string& ref;
    const std::string* ptr;

    std::string const& blaxxx;

    ::std::gar::function<void()> a;

    int field[100];
};
```

The problem also exists for Java and C#, but I have not fixed it since I am not an expert in these languages.

On a side-node: I would have loved to add the class above to the test sources, but I am not very happy with impacting all the other tests when modifying the test cpp sources, did you already consider separating tests, so that things can be verified in isolation?

Cheers

### CODE REVIEW

**Comment by avkonst** [comment]
"The problem also exists for Java and C#, but I have not fixed it since I am not an expert in these languages." 
> please create a ticket for C# and Java including as much context as you have got.

"I would have loved to add the class above to the test sources, but I am not very happy with impacting all the other tests when modifying the test cpp sources"
> You can do it the following way: make sure there are no code changes, run tests, make sure passed. Add new test sample code, run tests regenerating gold files, commit. Use new code sample by new tests.

"but I am not very happy with impacting all the other tests when modifying the test cpp sources"
> Sure, then separate.

"did you already consider separating tests, so that things can be verified in isolation?"
> you do not need to refactor existing tests, you can start with only your set of tests being separate. Feel free to separate it if you need.


In relation to the code change.
I trust you know what you are doing. Review of regular expressions is impossible job. I will do only a quick look.

**Comment by avkonst** [comment]
Once you add your proposed code sample with fields into tests (separated suite or existing), I will merge your request.
Thanks for contributing. Very impressive.

**Comment by noricor** [comment]
Hi,

thanks for feedback, I will give it a try and separate the tests, because it would give more confidence when
implementing such regex magic.

As proposed, I will also add the tickets for the open work.

There is also one related thing regarding the detection of function and variable declarations (not definitions)
in header files, here some improvements could enable us to do more powerful checking, ticket is coming soon ...

**Comment by avkonst** [comment]
Let me know when I can review again. I am off work for a weekend.

**Comment by noricor** [comment]
sure, while improving the solution I found that currently we will always have some false positives here,
so I will update the MR and we can see if it is still worth merging it in.

File: metrixpp/tests/general/test_std_code_member/sources/test.cpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
+#include "test.hpp"
+#include <iostream>
+
+// false positive: the parser cannot identify these 3 lines as a single statement
+   typedef typename boost::intrusive::
+      pointer_traits<void_pointer>::template
+         rebind_pointer<block_ctrl>::type                         block_ctrl_ptr;
```
Comment (noricor) [inline metrixpp/tests/general/test_std_code_member/sources/test.cpp:7]: Here I hit a problem, can this be solved on the parser level somehow?

Comment (avkonst) [reply metrixpp/tests/general/test_std_code_member/sources/test.cpp:7]: As per my comment here: https://github.com/metrixplusplus/metrixplusplus/issues/42#issuecomment-735456563
Can we just count `;` symbols to get number of fields within a class? It is certainly best way of finding god classes.

> "Here I hit a problem, can this be solved on the parser level somehow?"
This is detected as a field. Why would not you want to count it as a member? It is type defining member. And I consider well contributes to spotting large classes.

I guess the goal of the tool was to:
- allow spotting of `bad quality code` using `good enough` counting with simple implementation and fast performance Over
- precise counting for everything and poor performance. 

So, to answer your question - I do not think there is anything within the parser to differentiate different types of fields.

What are the other "false positives" left? I just realized that this mentioned false positive was not considered a false positive in my code bases when I applied the tool back in 2008-2013 years. I wanted to make classes not big regardless of of what made them big...

**Comment by noricor** [comment]
Hi, concluding the discussion above this does not need to be merged in, 
since we want to keep things simple and fast.

Nevertheless, the current implementation of the regular expression in std.code.member.fields 
remains not correct since it assumes that field names are exactly 2 characters long. 
So what to do with that metric?

**Comment by avkonst** [comment]
The original pull request description says:
The original problem was a missing * in the regular expression

```
r'''([_a-zA-Z][_a-zA-Z0-9]*\s+[_a-zA-Z][_a-zA-Z0-9])\s*[=;]''')
```
the expression should instead be

```
r'''([_a-zA-Z][_a-zA-Z0-9]*\s+[_a-zA-Z][_a-zA-Z0-9]*)\s*[=;]''')
```

You are saying this still does not detect many fields. Right?
If yes, is it possible to assume that number of data members/fields is the same as total number of nested members minus all other types of nested members (ie. classes, structs, functions, etc.. all but not variables). If this assumption is working assumption, I think it should be possible to count total number of members by searching for `;` occurrences in a region excluding its children regions. Next finding all function and operator members by searching for a kind of `(`/`)`, which indicates a member is a function. And next number of data members/fields should be about total number of members minus number of function members. What do you think?

**Pull request closed without merging**